plugins/SpectrumAnalyzer/SaProcessor.cpp

```diff
@@ -0,0 +1,556 @@
+		int start_time = std::chrono::high_resolution_clock::now().time_since_epoch().count();
+	#endif
+	// only take in data if any view is visible and not paused
+	if ((m_spectrumActive || m_waterfallActive) && !m_controls->m_pauseModel.value())
+	{
+		const bool stereo = m_controls->m_stereoModel.value();
+		fpp_t in_frame = 0;
+		while (in_frame < frame_count)
+		{
+			// fill sample buffers and check for zero input
+			bool block_empty = true;
+			for (; in_frame < frame_count && m_framesFilledUp < m_inBlockSize; in_frame++, m_framesFilledUp++)
+			{
+				if (stereo)
+				{
+					m_bufferL[m_framesFilledUp] = in_buffer[in_frame][0];
+					m_bufferR[m_framesFilledUp] = in_buffer[in_frame][1];
+				}
+				else
+				{
+					m_bufferL[m_framesFilledUp] = (in_buffer[in_frame][0] + in_buffer[in_frame][1]) * 0.5;
+					m_bufferR[m_framesFilledUp] = (in_buffer[in_frame][0] + in_buffer[in_frame][1]) * 0.5;
+				}
+				if (in_buffer[in_frame][0] != 0 || in_buffer[in_frame][1] != 0)
+				{
+					block_empty = false;
+				}
+			}
+	
+			// Run analysis only if buffers contain enough data.
+			// Also, to prevent audio interruption and a momentary GUI freeze,
+			// skip analysis if buffers are being reallocated.
+			if (m_framesFilledUp < m_inBlockSize || m_reallocating) {return;}
+
+			// update sample rate
+			m_sampleRate = Engine::mixer()->processingSampleRate();
+	
+			// apply FFT window
+			for (int i = 0; i < m_inBlockSize; i++)
+			{
+				m_bufferL[i] = m_bufferL[i] * m_fftWindow[i];
+				m_bufferR[i] = m_bufferR[i] * m_fftWindow[i];
+			}
+	
+			// lock data shared with SaSpectrumView and SaWaterfallView
+			QMutexLocker lock(&m_dataAccess);
+
+			// Run FFT on left channel, convert the result to absolute magnitude
+			// spectrum and normalize it.
+			fftwf_execute(m_fftPlanL);
+			absspec(m_spectrumL, m_absSpectrumL.data(), binCount());
+			normalize(m_absSpectrumL, m_normSpectrumL, m_inBlockSize);
+	
+			// repeat analysis for right channel if stereo processing is enabled
+			if (stereo)
+			{
+				fftwf_execute(m_fftPlanR);
+				absspec(m_spectrumR, m_absSpectrumR.data(), binCount());
+				normalize(m_absSpectrumR, m_normSpectrumR, m_inBlockSize);
+			}
+
+			// count empty lines so that empty history does not have to update
+			if (block_empty && m_waterfallNotEmpty)
+			{
+				m_waterfallNotEmpty -= 1;
+			}
+			else if (!block_empty)
+			{
+				m_waterfallNotEmpty = m_waterfallHeight + 2;
+			}
+
+			if (m_waterfallActive && m_waterfallNotEmpty)
+			{
+				// move waterfall history one line down and clear the top line
+				QRgb *pixel = (QRgb *)m_history.data();
+				std::copy(pixel,
+						  pixel + binCount() * m_waterfallHeight - binCount(),
+						  pixel + binCount());
+				memset(pixel, 0, binCount() * sizeof (QRgb));
+
+				// add newest result on top
+				int target;		// pixel being constructed
+				float accL = 0;	// accumulators for merging multiple bins
+				float accR = 0;
+	
+				for (int i = 0; i < binCount(); i++)
```

Comment:
Your compiler is set to _very paranoid_. xD
If it was up to me, I would rather use `uint8_t` etc., which is far more transparent than `std::size_t`. It seems a huge waste to use a 32 or 64 bit counters to count from 1 to 10..

Reply:
(But fixing signed / unsigned probably makes some sense, although all the for() lines will be 9 characters longer..)

Reply:
Yeah, I use `-Weverything` (with a few exceptions of really annoying warnings).

The thing is, a 32 bit architecture can not count with less than 32 bit, and a 64 bit architecture... I did benchmarks where using `char`s instead of `int`s increased runtime drastically. Especially in arrays: If you have a char array and write one char into it, a 32 bit architecture can only write 4 chars, so it first reads 4 chars from the array, then replaces your char, then writes back, which makes using chars in array slow (and dangerous in multi-threading).

Also, as soon as you have `char x; my_vector[x];`, `x` is converted to `std::size_t`, so the space is wasted again (plus an unnecessary conversion).

If you're converned about linelength, you can use the [goes-to-operator](https://stackoverflow.com/questions/1642028/what-is-the-operator-in-c): `int i = binCount(); while (i --> 0) {}`.

Reply:
OK, most of the sign conversions should be fixed. At least those cases where it made sense (mainly counters, array indices etc.).
There is a lot of float-to-int conversions with Qt, which has real data types on some functions to allow sub-pixel precision, but apparently many other functions just use integers. Also, their width is signed, so there is some signedness paranoia warnings left in that area.
Hopefully I didn't break anything in the process; these changes should not make any semantic difference, but there was a lot of them, so one never knows.. :)

Reply:
If you think it's OK, I'd be for resolving this conversation.

Reply:
I pretty much thought it was OK-ish before. :D I feel like this work did not lead to any real improvement, but at least it is now kind of "more correct" and less misleading for future maintainers. So I'm happy to hit resolve, if only so that I can turn off the warning-flood switch.. ^^;